PJRT: Profiling Output to JSON

cryptodeal opened the pull request:
Implements conversion of PJRT protobuf profiling output to JSON format. Mirrors the XLA implementation, but is much quicker to compile.

gwenzek commented on pjrt/third_party/re2/re2.cc:
```diff
@@ -0,0 +1,28 @@
+    return RE2::FullMatch(op_type_view, *kJaxOpTypeRegEx);
+  }
+
+  bool isNumber(const char *str, size_t len) {
```
this can be done in zig with std.fmt.parseInt

gwenzek commented on pjrt/third_party/re2/re2.cc:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  bool isJaxOpType(const char *op_type, size_t len) {
+    absl::string_view op_type_view(op_type, len);
```
not a big fan of special casing. Also this is used to strip information ouptutted by the compiler, so I think we don't want to do that.  

overall I think we would be better by not trying to guess the source framework based on the casing of the operator name, and drop this dependency on re2.

cryptodeal replied:
Looking into this bit; this is based directly on how XLA parses: https://github.com/openxla/xla/blob/cb5bdb601bdb352afaa804b5b4a8826734772e3b/xla/tsl/profiler/utils/tf_op_utils.cc#L129

I agree it would be best to get rid of the `re2` and `abseil-cpp` dependencies.

gwenzek replied:
I understand XLA is doing weird stuff, but I don't think we should reproduce that part. The goal is to parse the files generated while profiling ZML, not tensorflow or jax. Also I believe this way of handling the input is very brittle, because it can categorize ZML operators as tensorflow / jax and apply logic where it's not warranted.
I'd rather remove this logic altogether.

gwenzek replied:
I don't think this was addressed

cryptodeal replied:
Gotcha, I'll just remove this logic entirely then.

cryptodeal replied:
Sorry, somehow I missed the comment you'd left on this 4 days ago after I'd initially revised to remove the dependencies on `re2`/`abseil-cpp`.

gwenzek commented on pjrt/convert/utils/xplane_visitor.zig:
```diff
@@ -0,0 +1,293 @@
+        cb: fn (allocator: std.mem.Allocator, xstat: XStatVisitor, ctx: ?*anyopaque) std.mem.Allocator.Error!void,
+        ctx: ?*anyopaque,
+    ) !void {
+        for (self.stats_owner.stats.items) |*stat| {
```
I don't really like this visitor pattern. We use it in zml because we need to handle user defined struct, but here you know all the structs, so I'm not sure what you gain compared to writing the for loop every time.

cryptodeal replied:
Yea, that makes sense; I wrote it this way purely because that's how the XLA implementation handles it, but will rewrite as for loops.

gwenzek commented on pjrt/convert/utils/xplane_schema.zig:
```diff
@@ -0,0 +1,554 @@
+pub const kCustomPlanePrefix = "/device:CUSTOM:";
+
+pub const HostEventType = enum(u16) {
+    kUnknownHostEventType = 0,
```
those kPascalCase should be replace by snake_case.

alternatively instead of using different name in the enum and in the json,
 just use the human readable name in the enum by using the `@"ExecutorState::Process"` syntax.

gwenzek commented on pjrt/convert/utils/trace_utils.zig:
```diff
@@ -0,0 +1,14 @@
+// Constants used as trace_viewer PID (device_id in trace_events.proto).
+// PID 0 is unused.
+// Support up to 500 accelerator devices.
+pub const kFirstDeviceId = 1;
```
just inline this inside "trace_container.zig" also no need for the "k" prefix. and use snake_case

gwenzek commented on pjrt/convert/utils/math_utils.zig:
```diff
@@ -0,0 +1,9 @@
+const std = @import("std");
+
+pub fn nanoToPico(n: u128) u128 {
```
please keep those helpers next to where they are used, or inline them.

gwenzek reviewed (commented):
thanks for the work, can you share with me a json file produced by this implementation ?

cryptodeal commented:
Here's the JSON outputted from `open_llama3b.trace.pb` file you'd sent over for testing:
[open_llama3b.trace.json](https://github.com/user-attachments/files/17653546/open_llama3b.trace.json)

cryptodeal commented:
Modified `TraceConverter` so that it accepts a `[]const u8` as input (i.e. can pass encoded profiling protobuffer directly to the converter and only write the JSON to a file).

Removed deps on `re2`/`abseil-cpp` as we're now using `mvzr` for simple regex library. Additionally, no information is being stripped from the resulting JSON while parsing TF/Jax ops, it's simply used to correctly categorize stat types based on operation name at this point (feature parity w XLA).

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,223 @@
+        resources: std.AutoHashMapUnmanaged(u32, Resource) = .{},
+
+        pub fn mutableResource(self: *Device, allocator: std.mem.Allocator, resource_id: u32) !*Resource {
+            const entry = try self.resources.getOrPut(allocator, resource_id);
```
I don't really like this pattern because it default inits value. I'd rather you use `getOrPutValue` directly from the caller.

gwenzek commented on pjrt/convert/utils/xplane_visitor.zig:
```diff
@@ -0,0 +1,252 @@
+        return self.metadata.display_name != .Empty;
+    }
+
+    pub fn displayName(self: *const XEventVisitor) []const u8 {
```
you should return an optional here, and drop `hasDisplayName()`

gwenzek commented on pjrt/convert/trace_converter.zig:
```diff
@@ -0,0 +1,178 @@
+        const Entry = std.AutoHashMapUnmanaged(u32, TraceContainer.Device).Entry;
+        const pairs = try allocator.alloc(Entry, self.container.metadata.devices.count());
+        defer allocator.free(pairs);
+        var iter = self.container.metadata.devices.iterator();
```
if you really need the index, please convert the HashMap to ArrayHashMap, and then you can directly `std.mem.sort(devices.entries)`,  or call `devices.sort()` if you need to rebuild the index afterward.

gwenzek commented on pjrt/convert/trace_converter.zig:
```diff
@@ -0,0 +1,178 @@
+                    \\,"args":{
+                );
+                const ArgsEntry = std.StringHashMapUnmanaged([]const u8).Entry;
+                const sorted_args = try allocator.alloc(ArgsEntry, event.args.count());
```
same thing, I have a hard time justifying a hashmap here for args.

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,223 @@
+        return self;
+    }
+
+    pub fn mutableDevice(self: *TraceContainer, allocator: std.mem.Allocator, device_id: u32) !*Device {
```
please also replace by `getOrPutValue`

gwenzek commented on pjrt/convert/utils/xplane_schema.zig:
```diff
@@ -0,0 +1,300 @@
+const std = @import("std");
+const tf_op_utils = @import("tf_op_utils.zig");
+
+pub const HostEventType = enum(u16) {
```
please add a comment explaining the casing is important to match the contant of the protobuffer

gwenzek commented on pjrt/convert/utils/xplane_schema.zig:
```diff
@@ -0,0 +1,300 @@
+};
+
+pub const StatType = enum(u16) {
+    UnknownStatType = 0,
```
just use "unknown", the type system knows it's a StateType

cryptodeal replied:
When initializing `XPlaneVisitor`, we need to be able to convert from the string representation of the enum value to the actual zig `StatType` enum type. XLA has the string representation for unknown ops as `"UnknownStatType"`. 

For simplicity of the `StatType.fromString` logic, imo it's best to leave as is so we can purely leverage `std.meta.stringToEnum`.

gwenzek replied:
oh, didn't realize this could happen in protobuf too, I thought it was your fallback for unhandled protobuffer fields

cryptodeal commented on pjrt/convert/trace_converter.zig:
```diff
@@ -0,0 +1,178 @@
+            \\{"displayTimeUnit":"ns","metadata":{"highres-ticks":true},"traceEvents":[
+        );
+
+        // TODO: finish implementing
```
Oops, this is just one that I forgot to remove

cryptodeal commented on pjrt/convert/utils/xplane_visitor.zig:
```diff
@@ -0,0 +1,252 @@
+        return self.metadata.name.getSlice();
+    }
+
+    pub fn toString(self: *const XStatVisitor, allocator: std.mem.Allocator) ![]const u8 {
```
I went with `AnyWriter` instead, since that works nicely with the `ArrayList`/`std.fs.File` writers.

gwenzek replied:
I think this changes havn't been pushed

gwenzek replied:
also can you rewrite this function to directly take an xplane and an xstat instead of a "xstatevisitor" ?

Since this visitor doesn't have state, it's only used to "extend" the XState proto, which is better done by just writing function using plain data as input.

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -184,12 +224,17 @@ pub const TraceContainer = struct {
-        const entry = try self.metadata.devices.getOrPut(allocator, device_id);
-        if (!entry.found_existing) {
-            entry.value_ptr.* = .{};
+    pub fn getOrPutDevice(self: *TraceContainer, allocator: std.mem.Allocator, key: u32) !GetOrPutResult(u32, ?trace_events_proto.Device) {
```
there is a misunderstanding there. This is not what I asked for. 
I want to you to adapt the data structure to the algorithm.

* if you need to insert stuff by key, please use an `AutoArrayHashMap`, and `getOrPutValue`
* if you don't please use an `ArrayList(KV)` and use `.append`. 

The answer to this question probably depend on the thing you are storing. Maybe devices need to use a hash map and arguments an arraylist.

in any case, please explain for each container why you chose this one and not the other.

cryptodeal replied:
For any protobuf `map<KeyType, ValueType>`, protobuf encodes as a repeated message where the message type is a pair of `KeyType, ValueType`, which is why `zig-protobuf` handles as `ArrayList(KV)`. However, protobuf does not allow duplicate keys, so a better translation would probably be `AutoArrayHashMap`.

cryptodeal replied:
For `devices`, I've chosen `AutoArrayHashMap` as it enforces deduplication on the key, which is consistent with the protobuf map type/XLA implementation.

For `args`, I've chosen `StringArrayHashMap` as it enforces deduplication on the key, which is consistent with the protobuf map type/XLA implementation. Additionally, the `KV` pairs here are directly outputted as JSON Key/Value pairs; this ensures the resulting output is valid JSON (no duplicate keys).

For `resources`, I've chosen `AutoArrayHashMap` as it enforces deduplication on the key value, which is consistent with the protobuf map type/XLA implementation.

gwenzek reviewed (changes requested):
there is a misunderstanding there. This is not what I asked for.
I want to you to adapt the data structure to the algorithm.

if you need to insert stuff by key, please use an AutoArrayHashMap, and getOrPutValue
if you don't please use an ArrayList(KV) and use append.

The answer to this question probably depend on how is the data structure in the protobuf, and on how you need to represent it in the json.
If  you need to merge information from different part of the protobuffer then you need an ArrayHashMap, if you don't merge things then ArrayList is enough.

Maybe devices need to use a hash map and arguments an arraylist.
in any case, please explain for each container why you chose this one and not the other.

In general Zig std is pretty high quality. If you find yourself implementing helper around std it means you're doing something wrong. Before you were implementing "sort" on a HashMap, now you are implementing "getOrPutValue" on an ArrayList. This means you're using the wrong container for your algorithm. 

also I don't think my comments about the unwarranted regex logic of "isJax", "isTensorflow" was accounted for, yet the conversation was marked as resolved. Can you be more conservative when resolving discussions ?

cryptodeal commented:
> also I don't think my comments about the unwarranted regex logic of "isJax", "isTensorflow" was accounted for, yet the conversation was marked as resolved. Can you be more conservative when resolving discussions ?

Understood, I think I resolved as I'd read your initial comment as needing to remove the dependencies and prevent trimming potentially useful information when filtering op name by string, but I see what you mean now and will completely remove this logic.

cryptodeal commented:
Upon further reflection, I believe that `AutoArrayHashMap`/`StringArrayHashMap` are the best data structures for use here; namely, protobuf maps do not allow duplicate key entries, so this ensures deduplication works consistently with the XLA implementation while providing a simple means to sort the entries.

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,211 @@
+        const sort_by_ordinal = (device_id == host_threads_device_id);
+        var ordinal: u32 = 0;
+        for (plane.plane.lines.items) |*l| {
+            var line = xplane_visitor.XLineVisitor.init(plane, l);
```
here instead of creating a "XLineVisitor" then calling displayeId and displayName on it, can you create a `fn xLineDisplayId(plane, line)` and `fn xLineDisplayName(plane, line)` and call that ?

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,211 @@
+            const resource_id: u32 = @intCast(xline.displayId());
+            if (std.mem.eql(u8, xline.displayName(), xla_async_op_line_name)) continue;
+            for (xline.line.events.items) |*e| {
+                const xevent = xplane_visitor.XEventVisitor.init(xline.plane, xline.line, e);
```
same here with xline. This code is obfuscated, since "xplane" and "l" have become "xline.plane" and "xline.line". With less intermediary objects it would be easier to follow.

gwenzek commented on pjrt/convert/utils/xplane_visitor.zig:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    pub fn getEventType(self: *const XPlaneVisitor, event_metadata_id: i64) ?xplane_schema.HostEventType {
+        return self.event_type_by_id.get(event_metadata_id);
```
instead of returning an "?HostEventType" return HostEventType with .unknown for non found values.

gwenzek commented on pjrt/convert/utils/xplane_visitor.zig:
```diff
@@ -0,0 +1,234 @@
+const xplane_proto = @import("//tsl:xplane_proto");
+const xplane_schema = @import("xplane_schema.zig");
+
+pub const XEventMetadataVisitor = struct {
```
this struct is just a tuple, can you remove it ?

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,211 @@
+                    try event.args.put(allocator, xstat.name(), stat_str);
+                }
+
+                for (xevent.event.stats.items) |*s| {
```
here also there is obfuscation: xevent.event is just our original `e`. I don't think this "renaming"  helps

gwenzek reviewed (changes requested):
Thanks, it's improving. The core algorithm looks good now,
 but can you streamline the code by removing all the visitor struct that barely do anything ?

you would need some helper function that work directly on the protobuf object themselves instead of wrapping them in visitor firsts.

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,239 @@
+
+    pub const Trace = struct {
+        devices: std.AutoArrayHashMapUnmanaged(u32, Device) = .{},
+        trace_events: std.ArrayListUnmanaged(TraceEvent) = .{},
```
I don't think this field is used at all ?

* if it's not used: please remove it, and replace `metadata.devices` by just `devices`
* if it's used: why do you have both `events` and `metadata.trace_events` ?

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,239 @@
+
+pub const TraceContainer = struct {
+    metadata: Trace = .{},
+    events: std.ArrayListUnmanaged(*TraceEvent) = .{},
```
why is this storing pointers instead of actual structs ?

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,239 @@
+
+    fn getTraceViewerMaxEvents() !u64 {
+        const kMaxEvents = 1000000;
+        if (std.posix.getenv("TF_PROFILER_TRACE_VIEWER_MAX_EVENTS")) |max_events| {
```
ugh, we don't do this here. Please move this to an argument to fromXSpace, and if needed to a cli arg

gwenzek commented on pjrt/convert/trace_container.zig:
```diff
@@ -0,0 +1,239 @@
+        }
+        // sort the events according to start time.
+        // TODO: partial sort would improve performance.
+        std.mem.sort(*TraceEvent, self.events.items, {}, struct {
```
it feels a bit late to start dropping events. Maybe we should instead put a limit earlier on while reading events from the protobuffer. But for now it's ok to assume we can handle the input trace file. Ultimately the user control the size of the trace based on how long they run the model.

gwenzek commented on pjrt/convert/xplane_visitor.zig:
```diff
@@ -0,0 +1,53 @@
+    event_metadata_by_id: std.AutoHashMapUnmanaged(i64, *const xplane_proto.XEventMetadata) = .{},
+    stat_metadata_by_id: std.AutoHashMapUnmanaged(i64, *const xplane_proto.XStatMetadata) = .{},
+
+    const EventTypeGetter = *const fn ([]const u8) xplane_schema.HostEventType;
```
those Getter declaration aren't used anymore

gwenzek reviewed (approved):
Nice, thanks.

Sorry for all the back and forth, but I think it was worth it. We went down to 700 lines from the original 1500 lines implementation.

There is a final quick fix, and can you double check that you can still convert and load the example trace I shared with you earlier ?

cryptodeal commented:
Correct, the example trace shared previously still converts as expected: 
[open_llama3b.trace.json](https://github.com/user-attachments/files/17804155/open_llama3b.trace.json)


And agreed, the code is so much cleaner and easier to follow after having made the suggested changes (in hindsight, I should have initially gone for a less literal direct port of the XLA implementation haha)
